dataverse/orgsvc/C#/RemoveUnrefrencedQueues/RemoveUnrefrencedQueues/RemoveUnrefrencedQueues.cs

```diff
@@ -0,0 +1,207 @@
+
+            // This query will find all queues that are no-longer applicable by joining the activityparty table and the msdyn_originatingqueue table on the related objects
+            // and then filtering out the originatingqueue objects that still have a activity party in the to/cc/bcc list
+            var query = new QueryExpression("msdyn_originatingqueue");
+
+            // Add columns to query.ColumnSet
+            query.ColumnSet.AddColumns("msdyn_createdentityid", "msdyn_createdentitytype", "msdyn_queueid");
+
+            FilterExpression queues = new FilterExpression(LogicalOperator.Or);
+            FilterExpression createdEntities = new FilterExpression(LogicalOperator.Or);
+            if (activityParties.Entities.Count > 0)
+            {
+                foreach (var activityParty in activityParties.Entities)
+                {
+                    //Sender,To,CC,BCC
+                    int typemask = activityParty.GetAttributeValue<OptionSetValue>("participationtypemask").Value;
+
+                    if (typemask >= 1 && typemask <= 4 && activityParty.GetAttributeValue<EntityReference>("partyid").LogicalName == "queue")
+                    {
+                        tracingService.Trace("RemoveUnrefrencedQueues.GetQueuesToRemove: Applicable Queue: " + activityParty.GetAttributeValue<EntityReference>("partyid").Id.ToString());
+
+                        ConditionExpression condition = new ConditionExpression("msdyn_queueid", ConditionOperator.Equal, activityParty.GetAttributeValue<EntityReference>("partyid").Id);
+                        queues.Conditions.Add(condition);
+                    }
+                    if (typemask == 13)
+                    {
+                        tracingService.Trace("RemoveUnrefrencedQueues.GetQueuesToRemove: Applicable Entity: " + activityParty.GetAttributeValue<EntityReference>("partyid").Id.ToString());
+
+                        ConditionExpression condition = new ConditionExpression("msdyn_createdentityid", ConditionOperator.Equal, activityParty.GetAttributeValue<EntityReference>("partyid").Id.ToString());
+                        createdEntities.Conditions.Add(condition);
+                    }
+
+                }
+            }
+
+            // Put all the query conditions together
+            query.Criteria.AddFilter(queues);
+            query.Criteria.AddFilter(createdEntities);
+
+            tracingService.Trace("RemoveUnrefrencedQueues.GetQueuesToRemove: Executing Retrival");
+
+            return service.RetrieveMultiple(query);
+        }
+
+
+        /// <summary>
+        /// Execute method that is required by the IPlugin interface.
+        /// </summary>
+        /// <param name="serviceProvider">The service provider from which you can obtain the
+        /// tracing service, plug-in execution context, organization service, and more.</param>
+        public void Execute(IServiceProvider serviceProvider)
```

Comment:
I don't see any error handling in this code.

As a sample, it should demonstrate this best practice and throw an InvalidPluginExecutionException even in the event of an unexpected error.